File: Zelig/Zelig/CompileTime/CodeGenerator/LLVMIR/Type.cpp
#include "llvmheaders.h"
#include "Type.h"
#include "Type_impl.h"
#include "Module.h"
#include "Module_impl.h"
#include "Value.h"
#include "Value_impl.h"
#include <string>

//--//

NS_MICROSOFT
NS_ZELIG
NS_LLVM

//--//

std::map<std::string, Type_impl* > Type_impl::typeImpls;
std::map<llvm::Type* , Type_impl* > Type_impl::typeImpls_reverseLookupForLLVMTypes;

void Type_impl::privateInit( Module_impl* owner, std::string name, int sizeInBits )
{
    this->owner = owner;
    this->sizeInBits = sizeInBits;
    this->name = name;
    this->isValueType = true;
    this->hasHeader = false;
    this->isBoxed = false;
    this->underlyingBoxedType = nullptr;
    this->underlyingPointerType = nullptr;
}

Type_impl::Type_impl( Module_impl* owner, std::string name, int sizeInBits, llvm::Type* ty )
{
    privateInit( owner, name, sizeInBits );
    llvmType = ty;
    typeImpls_reverseLookupForLLVMTypes[ llvmType ] = this;
}

Type_impl::Type_impl( Module_impl* owner, std::string name, int sizeInBits )
{
    privateInit( owner, name, sizeInBits );
    llvm::Module* module = owner->GetLLVMObject( );

    assert( module->getTypeByName( name ) == nullptr && "Trying to override typeimpl!" );

    if( name == "System.Void" )
    {
        llvmType = llvm::Type::getVoidTy( module->getContext( ) );
    }
    //Basic LLVM types used to break basic value types circular ref.
    else if( name == "LLVM.System.Boolean" || name == "LLVM.System.Char" || name == "LLVM.System.SByte"
             || name == "LLVM.System.Byte" || name == "LLVM.System.Int16" || name == "LLVM.System.UInt16"
             || name == "LLVM.System.Int32" || name == "LLVM.System.UInt32" || name == "LLVM.System.Int64"
             || name == "LLVM.System.UInt64" )
    {
        llvmType = llvm::Type::getIntNTy( module->getContext( ), sizeInBits );
    }
    else if( name == "LLVM.System.Single" )
    {
        llvmType = llvm::Type::getFloatTy( module->getContext( ) );
    }
    else if( name == "LLVM.System.Double" )
    {
        llvmType = llvm::Type::getDoubleTy( module->getContext( ) );
    }
    else if( name == "LLVM.System.IntPtr" || name == "LLVM.System.UIntPtr" )
    {
        llvmType = llvm::Type::getInt8PtrTy( module->getContext( ) );
    }
    else
    {
        llvmType = StructType::create( module->getContext( ), this->name );
    }

    typeImpls_reverseLookupForLLVMTypes[ llvmType ] = this;
}

Type_impl* Type_impl::GetOrInsertTypeImpl( Module_impl* owner, std::string name, int sizeInBits, llvm::Type* ty )
{
    if( typeImpls.find( name ) == typeImpls.end( ) )
    {
        typeImpls[ name ] = new Type_impl( owner, name, sizeInBits, ty );
    }
    return typeImpls[ name ];
}

Type_impl* Type_impl::GetOrInsertTypeImpl( Module_impl* owner, std::string name, int sizeInBits )
{
    if( typeImpls.find( name ) == typeImpls.end( ) )
    {
        typeImpls[ name ] = new Type_impl( owner, name, sizeInBits );
    }
    return typeImpls[ name ];
}

Type_impl* Type_impl::GetTypeImpl( std::string name )
{
    if( typeImpls.find( name ) == typeImpls.end( ) ) return nullptr;
    return typeImpls[ name ];
}

Type_impl* Type_impl::GetTypeImpl( llvm::Type* ty )
{
    Type_impl* retval = typeImpls_reverseLookupForLLVMTypes[ ty ];

    if( retval == nullptr )
    {
        if( isa<PointerType>( ty ) && isa<StructType>( ty->getPointerElementType( ) ) )
        {
            return  typeImpls_reverseLookupForLLVMTypes[ ty->getPointerElementType( ) ];
        }
        printf( "Dump of type that caused the assert: \n" );
        ty->dump( );
        assert( 0 );
    }
    return retval;
}

unsigned Type_impl::GetSizeInBits( )
{
    return sizeInBits;
}

unsigned Type_impl::GetSizeInBitsForStorage( )
{
    if( isValueType ) return GetSizeInBits( );
    return owner->GetPointerSize( );
}

void Type_impl::AddField( unsigned offset, Type_impl* type, bool forceInline, string name )
{
    Type_field* f = new Type_field( name, type, offset, forceInline );

    //
    // Add the field in offset order, by moving all other pointer to fields
    //
    fields.resize( fields.size( ) + 1 );
    size_t i = 0;
    for( i = fields.size( ) - 1; i > 0; --i )
    {
        if( fields[ i - 1 ]->offset <= f->offset )
            break;

        fields[ i ] = fields[ i - 1 ];
    }
    fields[ i ] = f;
}

void Type_impl::Internal_AddTypeToStruct( int& idx, vector < llvm::Type* > &llvmFields, size_t& i )
{
    fields[ idx ]->finalIdx = llvmFields.size( );
    if( fields[ idx ]->forceInline )
    {
        llvmFields.push_back( fields[ idx ]->type->GetLLVMObject( ) );
        i = ( i - 1 ) + fields[ idx ]->type->GetSizeInBits( ) / 8;
    }
    else
    {
        llvmFields.push_back( fields[ idx ]->type->GetLLVMObjectForStorage( ) );
        i = ( i - 1 ) + fields[ idx ]->type->GetSizeInBitsForStorage( ) / 8;
    }
    idx++;
}

void Type_impl::SetupFields( )
{
    if( !isa<StructType>( GetLLVMObject( ) ) || fields.size( ) == 0 ) return;

    vector < llvm::Type* > llvmFields;

    int idx = 0;

    for( size_t i = 0; i < sizeInBits / 8; ++i )
    {
        if( idx < fields.size( ) && fields[ idx ]->offset == i )
        {
            Internal_AddTypeToStruct( idx, llvmFields, i );
        }
        else
        {
            llvmFields.push_back( Type_impl::GetOrInsertTypeImpl( owner, "System.Byte", 8 )->GetLLVMObject( ) );
        }
    }

    for( ; idx < fields.size( ); ++idx )
    {
        size_t i = 0;

        Internal_AddTypeToStruct( idx, llvmFields, i );
    }

    cast<StructType>( llvmType )->setBody( llvmFields, true );
}

void Type_impl::SetHasHeaderFlag( bool val )
{
    hasHeader = val;
}

void Type_impl::SetValueTypeFlag( bool val )
{
    isValueType = val;
}

void Type_impl::SetBoxedFlag( bool val )
{
    isBoxed = val;
}

bool Type_impl::IsBoxed( )
{
    return isBoxed;
}

bool Type_impl::IsValueType( )
{
    return isValueType;
}

std::string Type_impl::GetName( )
{
    return name;
}

void Type_impl::Dump( )
{
    printf( "Name:%s\n", name.c_str( ) );
    llvmType->dump( );
    printf( "\n" );
}

llvm::Type* Type_impl::GetLLVMObjectForStorage( )
{
    if( isValueType ) return GetLLVMObject( );
    return GetLLVMObject( )->getPointerTo( );
}

llvm::Type* Type_impl::GetLLVMObject( )
{
    return llvmType;
}

//*************************************************//
//*************************************************//
//*************************************************//

_Type::_Type( _Module^ module, Type_impl* impl ) : module( module )
{
    _pimpl = impl;
}

void _Type::AddField( unsigned offset, _Type^ type, bool forceInline, String^ name )
{
    const char* nameStr = (const char*)( Marshal::StringToHGlobalAnsi( name ) ).ToPointer( );

    _pimpl->AddField( offset, type->_pimpl, forceInline, nameStr );
}

void _Type::SetupFields( )
{
    _pimpl->SetupFields( );
}

bool _Type::IsInteger( )
{
    return _pimpl->GetLLVMObject( )->isIntegerTy( );
}

bool _Type::IsFloat( )
{
    return _pimpl->GetLLVMObject( )->isFloatTy( );
}

bool _Type::IsDouble( )
{
    return _pimpl->GetLLVMObject( )->isDoubleTy( );
}

bool _Type::IsFloatingPoint( )
{
    return IsFloat( ) || IsDouble( );
}

bool _Type::IsPointer( )
{
    return _pimpl->GetLLVMObjectForStorage( )->isPointerTy( );
}

bool _Type::IsStruct( )
{
    return _pimpl->GetLLVMObjectForStorage( )->isStructTy( );
}

bool _Type::IsPointerPointer( )
{
    return _pimpl->GetLLVMObject( )->isPointerTy( ) && _pimpl->GetLLVMObject( )->getPointerElementType( )->isPointerTy( );
}

bool _Type::IsValueType( )
{
    return _pimpl->IsValueType( );
}

bool _Type::__op_Equality( _Type^ b )
{
    return _pimpl == b->_pimpl;
}

_Type^ _Type::GetBTUnderlyingType( )
{
    return module->GetType( gcnew String( ( string( "LLVM." ) + _pimpl->GetName( ) ).c_str( ) ) );
}

int _Type::GetSizeInBits( )
{
    return _pimpl->GetSizeInBits( );
}

int _Type::GetSizeInBitsForStorage( )
{
    return _pimpl->GetSizeInBitsForStorage( );
}

void _Type::SetValueTypeFlag( bool val )
{
    _pimpl->SetValueTypeFlag( val );
}

void _Type::Dump( )
{
    _pimpl->Dump( );
}

//--//

NS_END
NS_END
NS_END

File: Zelig/Zelig/CompileTime/CodeGenerator/LLVMIR2/_Type.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using Llvm.NET;
using Llvm.NET.DebugInfo;
using Llvm.NET.Types;

namespace Microsoft.Zelig.LLVM
{
    public class _Type
    {
        private static readonly Dictionary< string, _Type > TypeImplMap = new Dictionary< string, _Type >( );
        private static readonly Dictionary< IntPtr, _Type > TypeImplsReverseLookupForLlvmTypes =
            new Dictionary< IntPtr, _Type >( );

        private Module LlvmModule => Module.LlvmModule;

        internal List< DebugMemberInfo > DiFields = new List< DebugMemberInfo >( );

        // REVIEW: Any reason these can't be consolidated into a single UnderlyingElementType?
        //         Not like you can have a boxed pointer...
        internal _Type UnderlyingBoxedType;
        internal _Type UnderlyingPointerType;

        public _Type( _Module module, string name, int sizeInBits, IDebugType<ITypeRef,DIType> typeRef )
        {
            Module = module;
            SizeInBits = sizeInBits;
            IsValueType = true;
            Name = name;
            DebugType = typeRef;
            TypeImplsReverseLookupForLlvmTypes[ DebugType.TypeHandle ] = this;
        }

        public _Type( _Module owner, string name, int sizeInBits )
        {
            SizeInBits = sizeInBits;
            IsValueType = true;
            Name = name;
            Module = owner;

            Debug.Assert( LlvmModule.GetTypeByName( name ) == null, "Trying to override _Type" );

            // basic types need to use DebugBasicType since the LLVM primitive types are all uniqued
            // and some of them would end up overlapping (i.e. System.UInt16 and Char are both i16 to
            // LLVM but need distinct DIBasicType instances to describe them for the debugger )
            switch( name )
            {
            case "System.Void":
                // DIType == null means void
                DebugType = Llvm.NET.DebugInfo.DebugType.Create( LlvmModule.Context.VoidType, ( DIType )null );
                break;

            case "LLVM.System.Char":
                DebugType = new DebugBasicType( LlvmModule.Context.GetIntType( ( uint )sizeInBits ), LlvmModule, name, DiTypeKind.UTF );
                break;

            case "LLVM.System.Boolean":
                DebugType = new DebugBasicType( LlvmModule.Context.GetIntType( ( uint )sizeInBits ), LlvmModule, name, DiTypeKind.Boolean );
                break;

            case "LLVM.System.SByte":
            case "LLVM.System.Int16":
            case "LLVM.System.Int32":
            case "LLVM.System.Int64":
                DebugType = new DebugBasicType( LlvmModule.Context.GetIntType( ( uint )sizeInBits ), LlvmModule, name, DiTypeKind.Signed );
                break;

            case "LLVM.System.Byte":
            case "LLVM.System.UInt16":
            case "LLVM.System.UInt32":
            case "LLVM.System.UInt64":
                DebugType = new DebugBasicType( LlvmModule.Context.GetIntType( ( uint )sizeInBits ), LlvmModule, name, DiTypeKind.Unsigned );
                break;

            case "LLVM.System.Single":
                DebugType = new DebugBasicType( LlvmModule.Context.FloatType, LlvmModule, name, DiTypeKind.Float );
                break;

            case "LLVM.System.Double":
                DebugType = new DebugBasicType( LlvmModule.Context.DoubleType, LlvmModule, name, DiTypeKind.Float );
                break;

            case "LLVM.System.IntPtr":
            case "LLVM.System.UIntPtr":
                // REVIEW: Shouldn't this be an integer value and not a pointer? 
                //         Standard  semantics for IntPtr is that of an integer
                //         that is the same size as a native pointer.
                DebugType = new DebugPointerType( LlvmModule.Context.Int8Type, LlvmModule, null);
                break;

            default:
                // All other types are created as structures the Debug type is a temporary.
                // Creation of concrete DIType deferred until SetupFields when full field
                // layout information is known.
                DebugType = new DebugStructType( LlvmModule, name, LlvmModule.DICompileUnit, name );
                break;
            }

            TypeImplsReverseLookupForLlvmTypes[ DebugType.TypeHandle ] = this;
        }

        public bool IsDouble => DebugType.IsDouble;

        public bool IsFloat => DebugType.IsFloat;

        public bool IsFloatingPoint => DebugType.IsFloatingPoint;

        public bool IsInteger => DebugType.IsInteger;

        public bool IsPointer => GetLLVMObjectForStorage( ).IsPointer;

        public bool IsPointerPointer => DebugType.IsPointerPointer;

        public bool IsStruct => GetLLVMObjectForStorage( ).IsStruct;

        public bool IsValueType { get; set; }

        public bool IsBoxed { get; set; }

        // REVIEW:
        //  This appears to be unused, it is set from the constructors
        //  but has no calls to the getter... 
        public bool HasHeader { get; set; }

        internal _Module Module { get; }

        internal DIType DIType => DebugType.DIType;

        internal List< TypeField > Fields { get; } = new List< TypeField >( );

        internal List< _Type > FunctionArgs { get; } = new List< _Type >( );

        public IDebugType<ITypeRef,DIType> DebugType { get; }

        public int SizeInBits { get; }

        public string Name { get; }

        internal static _Type GetOrInsertTypeImpl( _Module owner, string name, int sizeInBits, IDebugType<ITypeRef,DIType> ty )
        {
            if( !TypeImplMap.ContainsKey( name ) )
            {
                TypeImplMap[ name ] = new _Type( owner, name, sizeInBits, ty );
            }

            return TypeImplMap[ name ];
        }

        internal static _Type GetOrInsertTypeImpl( _Module owner, string name, int sizeInBits )
        {
            if( !TypeImplMap.ContainsKey( name ) )
            {
                TypeImplMap[ name ] = new _Type( owner, name, sizeInBits );
            }

            return TypeImplMap[ name ];
        }

        internal static _Type GetTypeImpl( string name )
        {
            _Type retVal;
            if( TypeImplMap.TryGetValue( name, out retVal ) )
                return retVal;

            return null;
        }

        internal static _Type GetTypeImpl( ITypeRef ty )
        {
            _Type retVal;
            if( TypeImplsReverseLookupForLlvmTypes.TryGetValue( ty.TypeHandle, out retVal ) )
                return retVal;

            var pointerType = ty as IPointerType;
            if( pointerType != null && pointerType.ElementType.IsStruct( ) )
            {
                return TypeImplsReverseLookupForLlvmTypes[ pointerType.ElementType.TypeHandle ];
            }

            Debug.Assert( false );
            return null;
        }

        public int GetSizeInBitsForStorage( )
        {
            if( IsValueType )
                return SizeInBits;

            return ( int ) ( LlvmModule.Layout.PointerSize( ) )*8;
        }

        public void AddField( uint offset, _Type type, bool forceInline, string name )
        {
            var f = new TypeField( name, type, offset, forceInline );

            // Add the field in offset order, by moving all other pointer to fields
            // Not the most efficient approach for C# but mirrors the original C++ code
            Fields.Add( default(TypeField) ); // increase count by one

            var i = 0;
            for( i = Fields.Count - 1; i > 0; --i )
            {
                if( Fields[ i - 1 ].Offset <= f.Offset )
                    break;

                Fields[ i ] = Fields[ i - 1 ];
            }
            Fields[ i ] = f;
        }

        public _Type GetBTUnderlyingType( ) => Module.GetType( $"LLVM.{Name}" );

        private void AddTypeToStruct( TypeField field, ICollection< ITypeRef > llvmFields, ref uint offset )
        {
            field.FinalIdx = ( uint ) llvmFields.Count;
            if( field.ForceInline )
            {
                llvmFields.Add( field.MemberType.DebugType );
                offset += ( uint ) field.MemberType.SizeInBits/8;
            }
            else
            {
                llvmFields.Add( field.MemberType.GetLLVMObjectForStorage( ) );
                offset += ( uint ) field.MemberType.GetSizeInBitsForStorage( )/8u;
            }
        }

        public void SetupFields( )
        {
            var structType = DebugType as DebugStructType;
            if( structType == null )
                return;

            var llvmFields = new List< ITypeRef >( );
            var idx = 0;
            uint offset = 0;
            while( offset < SizeInBits/8 )
            {
                if( ( idx < Fields.Count ) && ( Fields[ idx ].Offset == offset ) )
                {
                    AddTypeToStruct( Fields[ idx ], llvmFields, ref offset );
                    AddDiField( Fields[ idx ], idx, offset );
                    ++idx;
                }
                else
                {
                    // Add explicit padding if necessary.
                    // TODO: Clean this up with a single byte array [ n x i8 ]
                    llvmFields.Add( GetOrInsertTypeImpl( Module, "System.Byte", 8 ).DebugType );
                    ++offset;
                }
            }

            // Add in any remaining fields beyond the expected structure size. These should generally be zero-sized,
            // such as System.Object and the elements trailing strings/arrays.
            for( ; idx < Fields.Count; ++idx )
            {
                // NOTE: order of calls matters here as offset is a ref param and updated
                //       in the call to AddTypeToStruct
                AddDiField( Fields[ idx ], idx, offset );
                AddTypeToStruct( Fields[ idx ], llvmFields, ref offset );
            }

            // BUGBUG: We don't yet handle explicit struct layout with overlapping fields.

            structType.SetBody( true, LlvmModule, LlvmModule.DICompileUnit, null, 0, 0, llvmFields, DiFields, (uint)SizeInBits, 0 );
        }

        private void AddDiField( TypeField field, int index, uint offset )
        {
            var structType = DebugType as IStructType;
            Debug.Assert( structType != null );
            Debug.Assert( field.MemberType.DIType != null );
            var memberDiType = field.MemberType.GetDiTypeForStack( );
            if( field.ForceInline )
                memberDiType = field.MemberType.DebugType;

            // TODO: since Zelig takes explicit control of layout the alignment needs to come from Zelig somewhere...
            var debugMemberInfo = new DebugMemberInfo { Name = field.Name
                                                      , Type = memberDiType
                                                      , Index = (uint)index
                                                      , BitSize = (uint)field.MemberType.SizeInBits
                                                      , BitOffset = offset * 8UL
                                                      , BitAlignment = 0  /* ???? */
                                                      /* todo: file, line...*/
                                                      };
            DiFields.Add( debugMemberInfo );
        }

        internal IDebugType<ITypeRef, DIType> GetLLVMObjectForStorage( )
        {
            if( IsValueType )
                return DebugType;

            return new DebugPointerType( DebugType, LlvmModule );
        }

        internal IDebugType<ITypeRef,DIType> GetDiTypeForStack( )
        {
            if( IsValueType )
                return DebugType;

            return new DebugPointerType( DebugType, LlvmModule );
        }

        internal void Dump( )
        {
            Console.WriteLine( "Name:{0}\n", Name );
            Console.WriteLine( DebugType.ToString( ) );
            Console.WriteLine( );
        }
    }
}
